### DISCUSSION TITLE
Add custom icon with JS script

### DISCUSSION THREAD

**Discussion by michaellw** [eoureo/obsidian-runjs#30] [Codes request] [answered]
Hello, your plugin is awesome, I would like to know how to add a custom icon with js script, for example add icon on the left side of the tab bar, I tried commander plugin and it can only add on the right side!
![image](https://github.com/eoureo/obsidian-runjs/assets/276357/b659ba3f-1531-4474-9ccb-d3191aaf15cb)

**Comment by eoureo** [comment]
Hello.
Thank you for your interest.
RunJS still has a lot of room for improvement. I would love to hear your thoughts.


Upon considering your question, it seems that you can use the layout-change event of app.workspace to achieve what you want.

Upon examining the source code of Commander, it was indeed utilizing the layout-change event.

pageHeaderManager.ts - obsidian-commander [GitHub] - Visual Studio Code - GitHub  
https://github.dev/phibr0/obsidian-commander  

```js
class PageHeaderManager


init()

  this.plugin.registerEvent(
    app.workspace.on("layout-change", () => {
      this.addButtonsToAllLeaves();
    })


addPageHeaderButton()

  const buttonIcon = (view as ItemView). addAction(icon, name, () => {
    app.workspace.setActiveLeaf(leaf, { focus: true });
    app.commands.executeCommandById(id);
  });
```

In the above code, it uses itemView.addAction(Obsidian API) to place the icon in leaf's .view-header > .view-action. Hence, the icon is only placed in the designated .view-action (at the right).

To place the icon on the left side, you'll need to create code as Obsidian API doesn't have that functionality. Perhaps, you can place it in .view-header > .view-header-nav-buttons. Or after that element.

I think you should use appendChild to create a button like the addPageHeaderButton function.

I hope for your opinion.

**Reply by michaellw** [reply]
Hello, thanks for your reply, I'm actually not familiar with Obsidian API or coding, I'm limited to reading and writing simple code, so if you could give some examples that would be great!

**Reply by eoureo** [reply]
I'll post the code soon. Please wait a bit.

**Comment by eoureo** [comment, marked as answer]
Sorry for posting late.
Try the following and give me your comments.

1. Put the following codes into an appropriate note file.
2. Open RunJS settings.
3. Click the "+" button in Event handler to add code.
4. Set the event handler with worksapce - layout-change and activate it.

From now on, the command icon will refresh when you open or close another note file, or when a new tab is created or delete

````markdown

```js RunJS="{n:'Examples/Event/Add command icon'}"
import { ItemView, setIcon } from 'obsidian';

this.app.workspace.iterateAllLeaves((leaf) => {
    console.log("Add command icon");
    if (!(leaf.view instanceof ItemView)) return;		
    if (!(leaf.view?.headerEl)) return;
    
    const commandId = "my_icon";
    
    // to avoid duplication
    const olds = leaf.view.headerEl.querySelectorAll("#" + commandId);
    Array.from(olds).forEach(el => {
      el.remove();
    });
    
    const span = leaf.view.headerEl.createSpan({
            cls:"clickable-icon my_commanders",
            attr:{
              id: commandId,
              title:"My Command Icon",
              style:"margin-left: 8px; color: red"
            }
          });
    setIcon(span, "lucide-heart-handshake"); // Copy from icon list
    span.addEventListener("click", () => {
      // this.alert("hello!!"); // O.K.
      
      // run another code of RunJS
      // this.runCodeByName("Open with/VSCode"); // O.K.
      
      // run another codemmand of Obsidian
      // Replace id ""app:open-help"" with something else
      // To see the list of commands:
      //        Developer tools(Ctrl + Shift + I) - console
      //        input "app.commands.commands"
      this.app.commands.executeCommandById("app:open-help"); // O.K.
    });
    
    // change the insertion position
    leaf.view.forwardButtonEl.parentElement.appendChild(span);
  }
);
```

````

**Reply by michaellw** [reply]
Wow, this example is great, I think I know how to customize my commands now, thank you!

**Reply by eoureo** [reply]
I'm glad I could help.
Any questions or comments are welcome.
